cli/maintainer/src/commands_solido.rs

```diff
@@ -892,3 +875,100 @@ pub fn command_withdraw(
     };
     Ok(result)
 }
+
+#[derive(Serialize)]
+pub struct DeactivateValidatorIfCommissionExceedsMaxOutput {
+    // List of validators that exceeded max commission
+    entries: Vec<ValidatorViolationInfo>,
+    max_commission_percentage: u8,
+}
+
+#[derive(Serialize)]
+struct ValidatorViolationInfo {
+    #[serde(serialize_with = "serialize_b58")]
+    pub validator_vote_account: Pubkey,
+    pub commission: u8,
+}
+
+impl fmt::Display for DeactivateValidatorIfCommissionExceedsMaxOutput {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(
+            f,
+            "Maximum validation commission: {}",
+            self.max_commission_percentage
+        )?;
+
+        for entry in &self.entries {
+            writeln!(
+                f,
+                "Validator vote account: {}, validation commission: {}",
+                entry.validator_vote_account, entry.commission
+            )?;
+        }
+        Ok(())
+    }
+}
+
+/// CLI entry point to punish validator for commission violation.
+pub fn command_deactivate_validator_if_commission_exceeds_max(
+    config: &mut SnapshotConfig,
+    opts: &DeactivateValidatorIfCommissionExceedsMaxOpts,
+) -> solido_cli_common::Result<DeactivateValidatorIfCommissionExceedsMaxOutput> {
+    let solido = config.client.get_solido(opts.solido_address())?;
+
+    let mut violations = vec![];
+    for pubkey_entry in solido.validators.entries {
+        let validator = pubkey_entry.entry;
+        let vote_pubkey = pubkey_entry.pubkey;
+        let validator_account = config.client.get_account(&vote_pubkey)?;
+        let commission = get_vote_account_commission(&validator_account.data)
+            .ok_or_else(|| CliError::new("Validator account data too small"))?;
+
+        if !validator.active || commission <= solido.max_commission_percentage {
+            continue;
+        }
+
+        let instruction = lido::instruction::deactivate_validator_if_commission_exceeds_max(
+            opts.solido_program_id(),
+            &lido::instruction::DeactivateValidatorIfCommissionExceedsMaxMeta {
+                lido: *opts.solido_address(),
+                validator_vote_account_to_deactivate: vote_pubkey,
+            },
+        );
+        let signers: Vec<&dyn Signer> = vec![];
+        config.sign_and_send_transaction(&[instruction], &signers)?;
```

Comment:
Agree, that's a problem. Why do you think that it is not worth dealing with "we do need to be careful that the transaction can fit that many instructions"? Because it is improbable to have so many misbehaved validators?

Reply:
Yes exactly, the probability that we hit that case in production is extremely low because:

* There need to be multiple misbehaving validators at the same time, and even a single validator misbehaving should be very rare.
* Maintainers would already deactivate the misbehaving validator before we can run this, and we have 4 independent parties operating an instance.
* The instruction is small, it needs to reference only 3 accounts, so you can probably fit more than 3 of them in a single transaction.

And even if it does happen, it’s not the end of the world if we need to edit the code a bit and recompile and it takes a few more hours before those validators get deactivated. I would say it’s not worth complicating the code for right now.